From 24b176aabf3284e59e19269e77049ad091340624 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Thu, 20 Feb 2025 17:18:49 +0800
Subject: [PATCH] dxgi: Move the fallback device window to the bottom at
 creation.

Tidy Cauldron (2708320), a Unity game, uses EnumWindows() to find the first window in the current
thread and maximizes the window when changing to windowed mode. However, before this patch, the IME
UI window and the DXGI fallback device window are on top of the game window at creation and thus
they could get maximized instead. This causes the game window to lose focus and freeze.
---
 dlls/d3d11/tests/d3d11.c | 2 --
 dlls/dxgi/factory.c      | 1 +
 2 files changed, 1 insertion(+), 2 deletions(-)

diff --git a/dlls/d3d11/tests/d3d11.c b/dlls/d3d11/tests/d3d11.c
index d0031a81a7c..0a1dfa06160 100644
--- a/dlls/d3d11/tests/d3d11.c
+++ b/dlls/d3d11/tests/d3d11.c
@@ -2090,7 +2090,6 @@ static void test_create_device_child(void)
     EnumWindows(enum_first_current_thread_window_proc, (LPARAM)&result_hwnd);
     GetClassNameA(result_hwnd, class_name, ARRAY_SIZE(class_name));
     GetWindowTextA(result_hwnd, window_name, ARRAY_SIZE(window_name));
-    todo_wine
     ok(result_hwnd == hwnd, "Got unexpected window %p %s %s.\n", result_hwnd, window_name, class_name);
 
     ID3D11Device_Release(device);
@@ -2104,7 +2103,6 @@ static void test_create_device_child(void)
     EnumWindows(enum_first_current_thread_window_proc, (LPARAM)&result_hwnd);
     GetClassNameA(result_hwnd, class_name, ARRAY_SIZE(class_name));
     GetWindowTextA(result_hwnd, window_name, ARRAY_SIZE(window_name));
-    todo_wine
     ok(result_hwnd == hwnd, "Got unexpected window %p %s %s.\n", result_hwnd, window_name, class_name);
 
     ID3D11Device_Release(device);
diff --git a/dlls/dxgi/factory.c b/dlls/dxgi/factory.c
index 32d54f033eb..ebdb42a3806 100644
--- a/dlls/dxgi/factory.c
+++ b/dlls/dxgi/factory.c
@@ -617,6 +617,7 @@ HWND dxgi_factory_get_device_window(struct dxgi_factory *factory)
             ERR("Failed to create a window.\n");
             return NULL;
         }
+        SetWindowPos(factory->device_window, HWND_BOTTOM, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE | SWP_NOACTIVATE);
         TRACE("Created device window %p for factory %p.\n", factory->device_window, factory);
     }
 
-- 
GitLab

